docs/proposed/hs_reassembly.md

```diff
@@ -0,0 +1,410 @@
+
+## Goals
+
+1. Full backwards compatibility with the "development" branch of Mbed TLS,
+   unless the user explicitly enables fragmentation support in `config.h` and
+   via `mbedtls_ssl_conf_xxx()` API.
+2. Minimizing the conceptual changes to the Mbed TLS codebase.
+3. Robustness and simplicity.
+
+## Non-goals
+
+The below goals are out of scope of this PR, but will be addressed eventually.
+
+1. Minimal memory footprint when the fragmentation is enabled.
+2. Fragmenting the outgoing messages.
+
+# Detailed description of the problem
+
+Mbed TLS can only processs handshake messages that are fully contained within a
+single TLS record. If a handshake message spans multiple TLS records, Mbed TLS
+aborts the handshake.
+
+This ability is needed when the remote endpoint sends several concatenated handshake
+messages, e.g. `Server Hello || Server Certificate || Server Key Exchange ||
+Server Done`. If the length of the concatenated sequence exceeds the maximal
+length of a TLS record, the remote endpoint may chose to split a handshake
+message at the TLS record boundary.
+
+## Examples of fragmented handshake messages
+
+The simple scenario assumes that the fragmented handshake message starts and
+ends at the record boundary:
+
+```
++---------------------------+    +---------------------+
+|TLS                        |    |TLS                  |
+|hdr | payload              |    |hdr | payload        |
++----+----------------------+    +----+----------------+
+     | +------+------------+|         | +------------+ |
+     | |HS Hdr|Body Frag 0 ||  ...    | |Body Frag #n| |
+     | +------+------------+|         | +------------+ |
+     +----------------------+         +----------------+
+```
+
+More general scenario includes two TLS records (`m` and `m+1`), where:
+
+1. Record `m` contains the last fragment of handshake message `k-1`, followed by
+   the start of message `k`.
+2. Record `m+1` contains the end of message `k`, followed by start of message `k+1`.
+
+
+```
++--------------------------------------+  +-------------------------------------+
+|TLS msg #m                            |  |TLS  #m+1                            |
+|hdr | payload                         |  |hdr | payload                        |
++----+---------------------------------+  +----+--------------------------------+
+     |  +--------+-------------------+ |       | +------------+------+--------+ |
+     |  |.. #k-1 |#k ...             | |       | |... #k      | #k+1       ...| |
+     |  +--------+-------------------+ |       | +------------+------+--------+ |
+     +---------------------------------+       +--------------------------------+
+```
+
+# Proposed solution
+
+The suggested approach is using `mbedtls_reader` to accumulate the fragments of
+incoming handshake messages. Contents of TLS records that carry handshake data
+are appended to the accumulator, and the handshake state machine consumes
+messages from the accumulator:
```

Comment:
```suggestion
```suggestion
The library is using the `mbedtls_reader` structure to accumulate the handshake
fragments contained in a sequence of TLS records and presenting them as if they'd
reside in a contiguous buffer. 

Contents of TLS records that carry handshake data are fed to the reader, and the handshake
state machine consumes the handshake data as the granularity of handshake messages. The
necessary accumulation and reassembly is done within the reader.
```

Reply:
Addressed.